### PR TITLE
Sort config types by their load dependencies once

### PR DESCRIPTION
**Pull request by yhabteab**
This PR is basically the same as #10003, but since @Al2Klimov is out for this week, we need to get it done. So I've cherry-picked the commits from there and added some additional commits to address the requested changes in that PR.

- Loosely depends on https://github.com/Icinga/icinga2/pull/10169

closes #10003

### CODE REVIEW

File: test/base-type.cpp
Comment (julianbrost) [inline test/base-type.cpp:59]: That should probably use one of the check macros where you can specify a message with extra context. To find out what will happen, I added `load_after Service` to `host.ti`, and all I got was this not very helpful message:

```
error: in "base_type/sort_by_load_after": check true == previousTypes.find(dependency) != previousTypes.end() has failed [true != false]
```

**Review by julianbrost — CHANGES_REQUESTED** [review]
Compared to #10003, this PR has no cycle detection (at least I believe the purpose of some of the code might be cycle detection, it's not entirely clear though: https://github.com/Icinga/icinga2/pull/10003#discussion_r1746699570). When I intentionally introduced a dependency in the `*.ti` files, nothing in mkclass stopped me from doing so and all I got was a not very helpful error message in a test (see inline comment). So some more explicit "this failed due to a cycle" error reporting/`VERIFY()` might be a good idea.

File: test/base-type.cpp
Comment (julianbrost) [reply test/base-type.cpp]: Actually, as we just noticed, there are indeed not all libraries linked into the test binary. For example, if you introduce a cyclic dependency between `InfluxdbWriter` and `Influxdb2Writer`, the test currently won't catch it.

**Review by julianbrost — CHANGES_REQUESTED** [review]

File: test/CMakeLists.txt
Comment (julianbrost) [inline test/CMakeLists.txt:52]: There should at least be a comment what this does and why it's necessary. I'd say this is very non-obvious.

File: lib/base/type.hpp
Comment (julianbrost) [inline lib/base/type.hpp:96]: I wouldn't specify this here. It describes behavior that would mean that the function is broken and there's a test to check that the function isn't broken. So that shouldn't be relevant to any caller. A similar comment is also in the implementation (well, not of this function, but in the function that generates the result for this function). If you added this here because that might help with a failing test, I'd say a comment right above the assertion that would then fail would be more helpful.

**Review by julianbrost — CHANGES_REQUESTED** [review]

File: lib/base/type.hpp
Comment (julianbrost) [reply lib/base/type.hpp:96]: Note: the selected context is a bit too big. I **don't** want to say this for the following:

> Additionally, this will fail fatally when used prior to the completion of namespace initialization.

It's just regarding cycles. The caller doesn't provide any input that has to be free of cycles. If there were dependency cycles, this function would be broken and there's a tests to check that the function is broken.

To me, that part of the doc comment reads similar to "Note that this function wouldn't work if the implementation of std::sort() was broken"

**Review by julianbrost — COMMENTED** [review]
Looks fine and should do the job, I just don't want to do a quick approve + merge right before leaving for vacation and another pair of eyes shouldn't hurt. But I'd say you can optimistically adapt #10000 already.

I'm not 100% happy regarding the needs for the test changes including the CMakeLists changes, but that's probably what happens when relying on linker/loader magic for the type registration (the alternative would be to perform an extra self check within the implementation, so that it's checked every single time, but that's also not too nice).

File: lib/base/type.cpp
Comment (Al2Klimov) [inline lib/base/type.cpp:52]: The comment doesn't provide any benefit, does it?

File: lib/base/type.cpp
Comment (Al2Klimov) [inline lib/base/type.cpp:60]: * #10155

File: test/base-type.cpp
Comment (Al2Klimov) [inline test/base-type.cpp:75]: What about just comparing to all types|grep config type? 🙈

Comment (yhabteab) [reply test/base-type.cpp:75]: I am looking for config types that have to be always there, and unlike other types they cannot be disabled with cmake build flags like `-DICINGA2_WITH_ICINGADB=OFF`.

**Comment by yhabteab** [comment]
Oh! We can't backport this to 2.13.x due to https://github.com/Icinga/icinga2/commit/33e609d7919b79fb2ae7877ab367389045b4dec9.

**Comment by Al2Klimov** [comment]
Do you mean "We can't backport this to 2.13.x **without** https://github.com/Icinga/icinga2/commit/33e609d7919b79fb2ae7877ab367389045b4dec9"?

**Comment by yhabteab** [comment]
Just rebased it!

File: test/CMakeLists.txt
Comment (Al2Klimov) [inline test/CMakeLists.txt:57]: This is fine. 🎩🐶☕🔥🏠

**Review by Al2Klimov — APPROVED** [review]

File: test/base-type.cpp
Comment (julianbrost) [reply test/base-type.cpp:75]: > What about just comparing to all types|grep config type? 🙈

Do you mean actually spelling out all types here instead of the currently chosen subset? Or just filtering `Type::GetAllTypes()` and comparing that the same types are contained?

The first approach won't work properly as Yonas already said: the set of registered types is different based on build flags (and actually differs between our Linux and Windows builds).

The latter would work, and actually checking this additionally wouldn't be a bad idea, but some fixed "does it contain this specific type" checks should remain as this serves as a sanity check that some interesting types with dependencies were actually linked into the test binary. If the test binary was built without linking anything expect `lib/base/`, the set of registered types might be too small to contain anything useful as there's actually no `load_after` defined inside `lib/base/*.ti` and the test could actually become pretty pointless. So this is basically the defense against future changes to our CMakeLists accidentally rendering this test useless.

**Comment by yhabteab** [comment]
> Looks fine and should do the job, I just don't want to do a quick approve + merge right before leaving for vacation and another pair of eyes shouldn't hurt.

Since you're back now :) and the PR is still open, do you have any change requests? Otherwise, please also give your approval as well.

**Pull request merged**
